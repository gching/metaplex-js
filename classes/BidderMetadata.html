<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>BidderMetadata | metaplex-js</title>
	<meta name="description" content="Documentation for metaplex-js">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="../assets/css/main.css">
	<script async src="../assets/js/search.js" id="search-script"></script>
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.json" data-base="..">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="../index.html" class="title">metaplex-js</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="../modules.html">metaplex-js</a>
				</li>
				<li>
					<a href="BidderMetadata.html">BidderMetadata</a>
				</li>
			</ul>
			<h1>Class BidderMetadata</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<section class="tsd-panel tsd-hierarchy">
				<h3>Hierarchy</h3>
				<ul class="tsd-hierarchy">
					<li>
						<span class="target">BidderMetadata</span>
					</li>
				</ul>
			</section>
			<section class="tsd-panel-group tsd-index-group">
				<h2>Index</h2>
				<section class="tsd-panel tsd-index-panel">
					<div class="tsd-index-content">
						<section class="tsd-index-section ">
							<h3>Constructors</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-constructor tsd-parent-kind-class"><a href="BidderMetadata.html#constructor" class="tsd-kind-icon">constructor</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Properties</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#accountPublicKey" class="tsd-kind-icon">account<wbr>Public<wbr>Key</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#auctionPubkey" class="tsd-kind-icon">auction<wbr>Pubkey</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#bidderPubkey" class="tsd-kind-icon">bidder<wbr>Pubkey</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#cancelled" class="tsd-kind-icon">cancelled</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#lastBid" class="tsd-kind-icon">last<wbr>Bid</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class"><a href="BidderMetadata.html#lastBidTimestamp" class="tsd-kind-icon">last<wbr>Bid<wbr>Timestamp</a></li>
								<li class="tsd-kind-property tsd-parent-kind-class tsd-is-static"><a href="BidderMetadata.html#SCHEMA" class="tsd-kind-icon">SCHEMA</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Accessors</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-get-signature tsd-parent-kind-class"><a href="BidderMetadata.html#isCancelled" class="tsd-kind-icon">is<wbr>Cancelled</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Methods</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-method tsd-parent-kind-class tsd-is-static"><a href="BidderMetadata.html#decode" class="tsd-kind-icon">decode</a></li>
							</ul>
						</section>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Constructors</h2>
				<section class="tsd-panel tsd-member tsd-kind-constructor tsd-parent-kind-class">
					<a name="constructor" class="tsd-anchor"></a>
					<h3>constructor</h3>
					<ul class="tsd-signatures tsd-kind-constructor tsd-parent-kind-class">
						<li class="tsd-signature tsd-kind-icon">new <wbr>Bidder<wbr>Metadata<span class="tsd-signature-symbol">(</span>args<span class="tsd-signature-symbol">: </span><span class="tsd-signature-symbol">{ </span>accountPublicKey<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>auctionPubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>bidderPubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>cancelled<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">0</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">1</span><span class="tsd-signature-symbol">; </span>lastBid<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span><span class="tsd-signature-symbol">; </span>lastBidTimestamp<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span><span class="tsd-signature-symbol"> }</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><a href="BidderMetadata.html" class="tsd-signature-type" data-tsd-kind="Class">BidderMetadata</a></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L42">models/programs/auction/BidderMetadata.ts:42</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>args: <span class="tsd-signature-symbol">{ </span>accountPublicKey<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>auctionPubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>bidderPubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span><span class="tsd-signature-symbol">; </span>cancelled<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">0</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">1</span><span class="tsd-signature-symbol">; </span>lastBid<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span><span class="tsd-signature-symbol">; </span>lastBidTimestamp<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span><span class="tsd-signature-symbol"> }</span></h5>
									<ul class="tsd-parameters">
										<li class="tsd-parameter">
											<h5><span class="tsd-flag ts-flagOptional">Optional</span> account<wbr>Public<wbr>Key<span class="tsd-signature-symbol">?: </span><span class="tsd-signature-type">PublicKey</span></h5>
										</li>
										<li class="tsd-parameter">
											<h5>auction<wbr>Pubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span></h5>
										</li>
										<li class="tsd-parameter">
											<h5>bidder<wbr>Pubkey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">PublicKey</span></h5>
										</li>
										<li class="tsd-parameter">
											<h5>cancelled<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">0</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">1</span></h5>
										</li>
										<li class="tsd-parameter">
											<h5>last<wbr>Bid<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span></h5>
										</li>
										<li class="tsd-parameter">
											<h5>last<wbr>Bid<wbr>Timestamp<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">BN</span></h5>
										</li>
									</ul>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <a href="BidderMetadata.html" class="tsd-signature-type" data-tsd-kind="Class">BidderMetadata</a></h4>
						</li>
					</ul>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Properties</h2>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="accountPublicKey" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagOptional">Optional</span> account<wbr>Public<wbr>Key</h3>
					<div class="tsd-signature tsd-kind-icon">account<wbr>Public<wbr>Key<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">PublicKey</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L24">models/programs/auction/BidderMetadata.ts:24</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="auctionPubkey" class="tsd-anchor"></a>
					<h3>auction<wbr>Pubkey</h3>
					<div class="tsd-signature tsd-kind-icon">auction<wbr>Pubkey<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">PublicKey</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L28">models/programs/auction/BidderMetadata.ts:28</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="bidderPubkey" class="tsd-anchor"></a>
					<h3>bidder<wbr>Pubkey</h3>
					<div class="tsd-signature tsd-kind-icon">bidder<wbr>Pubkey<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">PublicKey</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L26">models/programs/auction/BidderMetadata.ts:26</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="cancelled" class="tsd-anchor"></a>
					<h3>cancelled</h3>
					<div class="tsd-signature tsd-kind-icon">cancelled<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">0</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">1</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L36">models/programs/auction/BidderMetadata.ts:36</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="lastBid" class="tsd-anchor"></a>
					<h3>last<wbr>Bid</h3>
					<div class="tsd-signature tsd-kind-icon">last<wbr>Bid<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">BN</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L30">models/programs/auction/BidderMetadata.ts:30</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class">
					<a name="lastBidTimestamp" class="tsd-anchor"></a>
					<h3>last<wbr>Bid<wbr>Timestamp</h3>
					<div class="tsd-signature tsd-kind-icon">last<wbr>Bid<wbr>Timestamp<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">BN</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L32">models/programs/auction/BidderMetadata.ts:32</a></li>
						</ul>
					</aside>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-property tsd-parent-kind-class tsd-is-static">
					<a name="SCHEMA" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagStatic">Static</span> SCHEMA</h3>
					<div class="tsd-signature tsd-kind-icon">SCHEMA<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">BorshSchema</span><span class="tsd-signature-symbol"> = ...</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L8">models/programs/auction/BidderMetadata.ts:8</a></li>
						</ul>
					</aside>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Accessors</h2>
				<section class="tsd-panel tsd-member tsd-kind-get-signature tsd-parent-kind-class">
					<a name="isCancelled" class="tsd-anchor"></a>
					<h3>is<wbr>Cancelled</h3>
					<ul class="tsd-signatures tsd-kind-get-signature tsd-parent-kind-class">
						<li class="tsd-signature tsd-kind-icon"><span class="tsd-signature-symbol">get</span> isCancelled<span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">boolean</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L38">models/programs/auction/BidderMetadata.ts:38</a></li>
								</ul>
							</aside>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">boolean</span></h4>
						</li>
					</ul>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Methods</h2>
				<section class="tsd-panel tsd-member tsd-kind-method tsd-parent-kind-class tsd-is-static">
					<a name="decode" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagStatic">Static</span> decode</h3>
					<ul class="tsd-signatures tsd-kind-method tsd-parent-kind-class tsd-is-static">
						<li class="tsd-signature tsd-kind-icon">decode<span class="tsd-signature-symbol">(</span>accountWithPubKey<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">AccountWithPublicKey</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><a href="BidderMetadata.html" class="tsd-signature-type" data-tsd-kind="Class">BidderMetadata</a></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/gching/metaplex-js/blob/d2c3ee6/src/models/programs/auction/BidderMetadata.ts#L58">models/programs/auction/BidderMetadata.ts:58</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>accountWithPubKey: <span class="tsd-signature-type">AccountWithPublicKey</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <a href="BidderMetadata.html" class="tsd-signature-type" data-tsd-kind="Class">BidderMetadata</a></h4>
						</li>
					</ul>
				</section>
			</section>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class=" ">
						<a href="../modules.html">Exports</a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
				</ul>
				<ul class="current">
					<li class="current tsd-kind-class">
						<a href="BidderMetadata.html" class="tsd-kind-icon">Bidder<wbr>Metadata</a>
						<ul>
							<li class=" tsd-kind-constructor tsd-parent-kind-class">
								<a href="BidderMetadata.html#constructor" class="tsd-kind-icon">constructor</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#accountPublicKey" class="tsd-kind-icon">account<wbr>Public<wbr>Key</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#auctionPubkey" class="tsd-kind-icon">auction<wbr>Pubkey</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#bidderPubkey" class="tsd-kind-icon">bidder<wbr>Pubkey</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#cancelled" class="tsd-kind-icon">cancelled</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#lastBid" class="tsd-kind-icon">last<wbr>Bid</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class">
								<a href="BidderMetadata.html#lastBidTimestamp" class="tsd-kind-icon">last<wbr>Bid<wbr>Timestamp</a>
							</li>
							<li class=" tsd-kind-property tsd-parent-kind-class tsd-is-static">
								<a href="BidderMetadata.html#SCHEMA" class="tsd-kind-icon">SCHEMA</a>
							</li>
							<li class=" tsd-kind-get-signature tsd-parent-kind-class">
								<a href="BidderMetadata.html#isCancelled" class="tsd-kind-icon">is<wbr>Cancelled</a>
							</li>
							<li class=" tsd-kind-method tsd-parent-kind-class tsd-is-static">
								<a href="BidderMetadata.html#decode" class="tsd-kind-icon">decode</a>
							</li>
						</ul>
					</li>
				</ul>
				<ul class="after-current">
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-constructor tsd-parent-kind-class"><span class="tsd-kind-icon">Constructor</span></li>
				<li class="tsd-kind-property tsd-parent-kind-class"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-class"><span class="tsd-kind-icon">Method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-class tsd-is-static"><span class="tsd-kind-icon">Static property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-class tsd-is-static"><span class="tsd-kind-icon">Static method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-interface"><span class="tsd-kind-icon">Property</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="../assets/js/main.js"></script>
</body>
</html>