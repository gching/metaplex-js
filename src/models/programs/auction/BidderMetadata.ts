import { PublicKey } from '@solana/web3.js';
import BN from 'bn.js';
import { deserializeUnchecked } from 'borsh';
import { BorshSchema } from '../../utils/types';
import { AccountWithPublicKey } from '../../../utils/privateTypes';

export class BidderMetadata {
  static SCHEMA: BorshSchema = new Map([
    [
      BidderMetadata,
      {
        kind: 'struct',
        fields: [
          ['bidderPubkey', 'pubkey'],
          ['auctionPubkey', 'pubkey'],
          ['lastBid', 'u64'],
          ['lastBidTimestamp', 'u64'],
          ['cancelled', 'u8'],
        ],
      },
    ],
  ]);

  accountPublicKey?: PublicKey;
  // Relationship with the bidder who's metadata this covers.
  bidderPubkey: PublicKey;
  // Relationship with the auction this bid was placed on.
  auctionPubkey: PublicKey;
  // Amount that the user bid.
  lastBid: BN;
  // Tracks the last time this user bid.
  lastBidTimestamp: BN;
  // Whether the last bid the user made was cancelled. This should also be enough to know if the
  // user is a winner, as if cancelled it implies previous bids were also cancelled.
  // 0 is false, 1 is true
  cancelled: 0 | 1;

  get isCancelled(): boolean {
    return this.cancelled === 1;
  }

  constructor(args: {
    accountPublicKey?: PublicKey;
    bidderPubkey: PublicKey;
    auctionPubkey: PublicKey;
    lastBid: BN;
    lastBidTimestamp: BN;
    cancelled: 0 | 1;
  }) {
    this.accountPublicKey = args.accountPublicKey;
    this.bidderPubkey = args.bidderPubkey;
    this.auctionPubkey = args.auctionPubkey;
    this.lastBid = args.lastBid;
    this.lastBidTimestamp = args.lastBidTimestamp;
    this.cancelled = args.cancelled;
  }

  static decode(accountWithPubKey: AccountWithPublicKey): BidderMetadata {
    const bidder = deserializeUnchecked(
      BidderMetadata.SCHEMA,
      BidderMetadata,
      accountWithPubKey.account.data,
    ) as BidderMetadata;

    return new BidderMetadata({
      ...bidder,
      accountPublicKey: accountWithPubKey.pubkey,
    });
  }
}
