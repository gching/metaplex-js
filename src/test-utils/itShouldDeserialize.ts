import { PublicKey } from '@solana/web3.js';
import { serialize } from 'borsh';
import { BorshSchema } from '../models/utils/types';

export function itShouldDeserialize<E>(
  cls: {
    SCHEMA: BorshSchema;
    // eslint-disable-next-line @typescript-eslint/ban-types
    decode: Function;
  },
  received: E,
  accountPubKey = PublicKey.default,
): void {
  const data = Buffer.from(serialize(cls.SCHEMA, received));

  const deserialized = cls.decode({
    pubkey: accountPubKey,
    account: {
      data,
    },
  }) as E;

  expect(deserialized).toDeepEqual(received);
}
