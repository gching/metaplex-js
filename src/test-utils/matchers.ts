import { isEqualWith } from 'lodash';
import BN from 'bn.js';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    interface Matchers<R> {
      toDeepEqual: <E = any>(expected: E) => R;
    }
  }
}

expect.extend({
  toDeepEqual<E = any>(received: E, expected: E) {
    const pass = isEqualWith(received, expected, (objValue, otherValue) => {
      if (BN.isBN(objValue) && BN.isBN(otherValue)) {
        return objValue.eq(otherValue);
      }
      return undefined;
    });

    if (pass) {
      return {
        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
        message: () => `expected ${received} not to deep equal ${expected}`,
        pass: true,
      };
    }

    return {
      // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
      message: () => `expected ${received} to deep equal ${expected}`,
      pass: false,
    };
  },
});

export default undefined;
