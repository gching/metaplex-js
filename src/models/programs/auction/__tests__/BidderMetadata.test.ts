import BN from 'bn.js';
import { PublicKey } from '@solana/web3.js';
import { BidderMetadata } from '../BidderMetadata';
import { itShouldDeserialize } from '../../../../test-utils/itShouldDeserialize';

describe('BidderMetadata', () => {
  it('should instantiate with the correct values', () => {
    const bidderMetadata = new BidderMetadata({
      bidderPubkey: PublicKey.default,
      auctionPubkey: PublicKey.default,
      lastBid: new BN(0),
      cancelled: 0,
      lastBidTimestamp: new BN(0),
    });

    expect(bidderMetadata).toMatchSnapshot();
  });

  it('should deserialize with the correct values', () => {
    const bidderMetadata = new BidderMetadata({
      bidderPubkey: PublicKey.default,
      auctionPubkey: PublicKey.default,
      lastBid: new BN(0),
      cancelled: 1,
      lastBidTimestamp: new BN(0),
    });

    bidderMetadata.accountPublicKey = PublicKey.default;

    itShouldDeserialize(BidderMetadata, bidderMetadata);
  });
});
