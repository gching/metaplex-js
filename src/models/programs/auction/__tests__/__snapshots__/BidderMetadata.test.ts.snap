// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidderMetadata should instantiate with the correct values 1`] = `
BidderMetadata {
  "accountPublicKey": undefined,
  "auctionPubkey": PublicKey {
    "_bn": "00",
  },
  "bidderPubkey": PublicKey {
    "_bn": "00",
  },
  "cancelled": 0,
  "lastBid": "00",
  "lastBidTimestamp": "00",
}
`;
